### PR TITLE
[Widget enhancement] 12 - Improve JSON-RPC bridging between iFrame and provider

### PR DESCRIPTION
**Pull request by anxolin**
# Summary

Part of the series https://github.com/cowprotocol/cowswap/pull/3812

When implementing https://github.com/cowprotocol/cowswap/pull/3822 I noticed a few improvements we could do in the RPC bridge.

This PR renames the component `JsonRpcManager` into `IframeRpcProviderBridge` (as i find it more intuitive what it does

Additionally, its code was enhanced to be more readable (i hope). It adds comments, more private functions to encapsulate some behaviour, and gives names to the props and methods that would make it easier to reason about.

Also, it adds a `disconnect` function, and it makes sure that when you connect a provider it correctly de-register the previous one and its events.  Before, I believe it was just nulling the provider. 

Also, I make sure that when you `connect` and you had a previous connection, then it will disconnect from the previous one before you connect the new one. 

I did the same for `disconnect`, now, it will de-register the listeners. 
`EDIT`: I had to undo the de-registering as, unfortunately in eip-1193 there's no `off` method: see 
https://github.com/cowprotocol/cowswap/pull/3824/commits/802f822a2579efdc4add91279c37659087ae061e

### CODE REVIEW

File: libs/widget-lib/src/IframeRpcProviderBridge.ts [hunk 1]
```diff
@@ -0,0 +1,126 @@
+import { EthereumProvider, JsonRpcRequest } from './types'
+
+const JSON_PRC_V = '2.0'
+const TARGET_ORIGIN = '*'
```
Comment (alfetopito) [inline libs/widget-lib/src/IframeRpcProviderBridge.ts:4]: TODO?

File: libs/widget-lib/src/IframeRpcProviderBridge.ts [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+      }
+      return
+    }
+
+    // Delegate the rpc call
+    this.processRequest(event.data)
+  }
+
+  private isRpcCall(event: MessageEvent) {
+    return event.data.jsonrpc === '2.0'
```
Comment (alfetopito) [inline libs/widget-lib/src/IframeRpcProviderBridge.ts:107]: Is that what the const is for?

```suggestion
    return event.data.jsonrpc === JSON_PRC_V
```

File: libs/widget-lib/src/IframeRpcProviderBridge.ts (hunk 1 above)
Comment (anxolin) [reply libs/widget-lib/src/IframeRpcProviderBridge.ts:4]: https://github.com/cowprotocol/cowswap/issues/3828

File: libs/widget-lib/src/IframeRpcProviderBridge.ts (hunk 2 above)
Comment (anxolin) [reply libs/widget-lib/src/IframeRpcProviderBridge.ts:107]: a339964e4f9c55495802af3aa1278ef8e79f26ba

Comment (shoom3301) [inline libs/widget-lib/src/IframeRpcProviderBridge.ts:107]: The method doesn't require `this`, so, would be better to declare it as a pure function with a return type

**Review by shoom3301 — APPROVED** [review]

File: libs/widget-lib/src/IframeRpcProviderBridge.ts (hunk 2 above)
Comment (anxolin) [reply libs/widget-lib/src/IframeRpcProviderBridge.ts:107]: 79799312f8f9df4ff6af6d609388c408f69c9c0e

**Pull request merged**